File: agent-driven-architecture-main/examples/typescript/src/blocks/escalation/tools.ts
// ── blocks/escalation/tools — the Verb table, with the gated verb ──────────
// `confirmEscalation` is `Irreversible`, and `Irreversible` CANNOT BE
// CONSTRUCTED without saying where its matching Request lives. That is 14.3's
// default-deny made structural: there is no default classification to forget,
// and the classification is a reviewed row in the registry rather than a guess.
//
// `requestedBy` reads the authority that ASKED straight off committed State —
// and it reads it from `TicketStatus`'s PARENT-DECLARED property, so it needs no
// match over the status variants and adding a fifth variant costs it nothing.
// The boundary compares that authority against the confirming one; this file
// never sees a Signature and never decides anything.

import type { Verb } from "@adr/spine/pure/verb";
import { irreversible, reversible } from "@adr/spine/pure/verb";
import { object, string } from "valibot";
import type {
  ConfirmEscalationCommand,
  ConfirmEscalationResult,
  RequestEscalationCommand,
  RequestEscalationResult,
} from "./contract";
import type { EscalationSlice } from "./slice";
import { statusOf } from "./slice";

const ticketInput = object({ ticket: string() });

export function escalationVerbs<S>(read: (state: S) => EscalationSlice): readonly Verb<S>[] {
  return [
    reversible<S, { ticket: string }, RequestEscalationResult, RequestEscalationCommand>({
      name: "requestEscalation",
      describe: "Request escalation of a ticket. Reversible; does NOT page on-call.",
      schema: ticketInput,
      run: (input) => ({ outcome: "ok", tool: "requestEscalation", ticket: input.ticket }),
      sign: (result, sig, id) => ({
        outcome: "ok",
        tool: "requestEscalation",
        sig,
        id,
        ticket: result.ticket,
      }),
    }),
    irreversible<S, { ticket: string }, ConfirmEscalationResult, ConfirmEscalationCommand>({
      name: "confirmEscalation",
      describe: "Confirm a pending escalation. IRREVERSIBLE: this pages on-call.",
      schema: ticketInput,
      run: (input) => ({ outcome: "ok", tool: "confirmEscalation", ticket: input.ticket }),
      sign: (result, sig, id) => ({
        outcome: "ok",
        tool: "confirmEscalation",
        sig,
        id,
        ticket: result.ticket,
      }),
      requestedBy: (state, result) => statusOf(read(state), result.ticket)?.requestedBy ?? null,
    }),
  ];
}

File: agent-driven-architecture-main/examples/kotlin/block/escalation/src/main/kotlin/adr/blocks/escalation/Tools.kt
// ── blocks/escalation/tools — the Verb table, including the gated verb ─────
// `confirmEscalation` is Verb.Irreversible, and Irreversible CANNOT BE CONSTRUCTED
// without `requestedBy`: the lens that reads, out of committed State, WHICH
// AUTHORITY asked for this. That lens is the whole of 14.3's "a different actor
// than the one that issued the Request", and the gate — not this file — enforces it.
//
// Note what this file cannot say: it never names Actor, Authority or Signature
// (gate check C4). A tool cannot ask who is asking, because the answer is stamped
// after it returns.
//
// THIS is the block where the split rule earns its keep. `lens` is read by the
// `requestedBy` row below and is fixed for the whole registration, so it becomes
// constructor state and drops out of the table — the same move the boundary makes with
// its registry in spine/boundary/action. The other five blocks take the same parameter
// for a uniform root, but this one is the reason the parameter exists.

package adr.blocks.escalation

import adr.contract.EscalationCommand
import adr.contract.EscalationResult
import adr.spine.pure.Lens
import adr.spine.pure.RawInput
import adr.spine.pure.TicketId
import adr.spine.pure.ToolName
import adr.spine.pure.Verb

val REQUEST_ESCALATION = ToolName("requestEscalation")
val CONFIRM_ESCALATION = ToolName("confirmEscalation")

internal data class TicketInput(val ticket: TicketId)

internal class EscalationTools<S>(private val lens: Lens<S, EscalationSlice>) {

    fun verbs(): List<Verb<S, *, *>> = listOf(
        Verb.Reversible(
            name = REQUEST_ESCALATION,
            describe = "Request escalation of a ticket. Reversible; does NOT page on-call.",
            decode = ::decodeTicket,
            run = { input, _ -> EscalationResult.RequestEscalation(REQUEST_ESCALATION, input.ticket) },
            sign = { r, sig, id -> EscalationCommand.RequestEscalation(r.tool, sig, id, r.ticket) },
            narrow = { it as? EscalationResult.RequestEscalation },
        ),
        Verb.Irreversible(
            name = CONFIRM_ESCALATION,
            describe = "Confirm a pending escalation. IRREVERSIBLE: it pages the on-call engineer.",
            decode = ::decodeTicket,
            run = { input, _ -> EscalationResult.ConfirmEscalation(CONFIRM_ESCALATION, input.ticket) },
            sign = { r, sig, id -> EscalationCommand.ConfirmEscalation(r.tool, sig, id, r.ticket) },
            narrow = { it as? EscalationResult.ConfirmEscalation },
            requestedBy = { state, result ->
                (result as? EscalationResult.ConfirmEscalation)
                    ?.let { lens(state).statusOf(it.ticket) }
                    ?.let { it as? TicketStatus.Escalating }
                    ?.requestedBy
            },
        ),
    )

    private fun decodeTicket(raw: RawInput): TicketInput? =
        raw.text("ticket")?.let { TicketInput(TicketId(it)) }
}
